design/design-generic-references.md

```diff
@@ -0,0 +1,197 @@
+the referenced resource and how to extract the value during development time of
+the provider. There are several cases where this is not possible:
+
+* The referencer field may not always target a single kind. For example, a URL
+  field in Route53 could get populated by a URL of an S3 bucket or access endpoint
+  of an RDS Instance.
+* Instead of hard-coding region for all managed resources, you may want to refer
+  to a `ConfigMap` or a `Secret` and let the managed resource take that
+  information from there.
+* In Composition, you can use values of the fields only between the children of
+  the same composite instance, but there are cases where you'd like to use a
+  separate resource that is not bound to any composite instance to store the
+  information and serve all composites.
+
+In order to enable such use cases, we need a generic version of the cross-resource
+references where you can specify all metadata required to find the referenced
+resource as well as details about how to extract the value you need and where to
+put it in the referencer object.
+
+## Proposal
+
+All managed resources and composite resources will have a top-level spec field
+that lists the generic references. They will be resolved in order and failure of
+any will cause reconciliation to fail. Inspired from the
+[latest developments in API conventions](https://github.com/kubernetes/community/pull/5748)
+regarding multi-kind references, the API will look like the following:
+```yaml
+# A generic reference to another managed resource.
+spec:
+  externalValues:
+  - fromObject:
+      group: ec2.aws.crossplane.io
+      version: v1alpha1
+      resource: vpcs
+      name: main-vpc
+      fieldPath: spec.forProvider.cidrBlock
+    toFieldPath: spec.cidrBlock
+```
+```yaml
+# A generic reference to a namespaced Kubernetes resource.
+spec:
+  externalValues:
+  - fromObject:
+      version: v1
+      resource: configmaps
+      name: common-settings
+      namespace: crossplane-system
+      fieldPath: data.region
+    toFieldPath: spec.forProvider.region
+```
+
+The syntax of `fieldPath` and `toFieldPath` fields will be similar to what we
+use in Composition, in line with [Kubernetes API Conventions](https://github.com/kubernetes/community/blob/744e270/contributors/devel/sig-architecture/api-conventions.md#selecting-fields).
+
+The provider pods will resolve the references via managed reconciler, hence they
+will need at least read-only permissions for any resource that's been referenced
+by an instance. Currently, every provider [has read access](https://github.com/crossplane/crossplane/blob/d8f57a8/internal/controller/rbac/provider/roles/roles.go#L66)
+to all of its own kinds, `Secret`s, `ConfigMap`s, `Lease`s and `Event`s in all
+namespaces. In order to reference resources other than the ones listed, users will
+have to create necessary RBAC resources manually to grant the permissions.
+
+### Implementation
```

Comment:
> It is possible we just say "you can do this, but you're going to likely have a bad time"

I think that'd be a good decision because I'm sure that people will have legitimate needs for such possible things that we cannot think of right now. If it's possibly hazardous but also worth doing for them, I wouldn't want to stand in the way.